pkg/monitor/controller.go

```diff
@@ -146,12 +178,44 @@ func (monitor *controllerMonitor) deleteStaleAgentCRDs() {
 		klog.Errorf("Failed to list agent monitoring CRDs: %v", err)
 		return
 	}
-	// Delete stale agent monitoring CRD based on existing nodes.
-	nodeLister := monitor.nodeInformer.Lister()
+	// Delete stale agent monitoring CRD based on existing Nodes and ExternalNodes.
+	externalNodeNames := sets.NewString()
+	if features.DefaultFeatureGate.Enabled(features.ExternalNode) {
+		externalNodes, err := monitor.externalNodeLister.List(labels.Everything())
+		if err != nil {
+			klog.Errorf("Failed to list ExternalNode CRDs: %v", err)
+			return
+		}
+		for _, en := range externalNodes {
+			externalNodeNames.Insert(en.Name)
+		}
+	}
 	for _, crd := range crds.Items {
-		_, err := nodeLister.Get(crd.Name)
-		if errors.IsNotFound(err) {
-			monitor.deleteAgentCRD(crd.Name)
+		_, err := monitor.nodeLister.Get(crd.Name)
+		if err != nil && errors.IsNotFound(err) {
+			if features.DefaultFeatureGate.Enabled(features.ExternalNode) {
+				if !externalNodeNames.Has(crd.Name) {
+					monitor.deleteAgentCRD(crd.Name)
+				}
+			} else {
+				monitor.deleteAgentCRD(crd.Name)
+			}
+		}
+	}
+
+}
+
+func (monitor *controllerMonitor) createAgentCRD(obj interface{}) {
+	node := obj.(*corev1.Node)
+	klog.Infof("Creating agent monitoring CRD %s", node.Name)
+	agentCRD := new(v1beta1.AntreaAgentInfo)
```

Comment:
It seems the logic is similar to that in function `createExternalNodeAgentCRD` from line 243-253. Is it possible to use a unique function to create AgentInfo, and give the name of Node/ExternalNode as parameter?